[mlir][memref] Add a new `ReifyResultShapes` pass

nicolasvasilache opened the pull request:
This pass reifies the shapes of a subset of `ReifyRankedShapedTypeOpInterface` ops with `tensor` results.

The pass currently only supports result shape type reification for:
   - tensor::PadOp
   - tensor::ConcatOp

It addresses a representation gap where implicit op semantics are needed to infer static result types from dynamic 
operands. But it does so by using `ReifyRankedShapedTypeOpInterface` as the source of truth rather than the op itself. 
As a consequence, this cannot generalize today.

TODO: in the future, we should consider coupling this information with op "transfer functions" (e.g. 
`IndexingMapOpInterface`) to provide a source of truth that can work across result shape inference, canonicalization and 
op verifiers.

The pass replaces the operations with their reified versions, when more static information can be derived, and inserts 
casts when results shapes are updated.

Example:
```mlir
  #map = affine_map<(d0) -> (-d0 + 256)>
  func.func @func(%arg0: f32, %arg1: index, %arg2: tensor<64x?x64xf32>) -> tensor<1x?x64xf32> {
    %0 = affine.apply #map(%arg1)
    %extracted_slice = tensor.extract_slice %arg2[0, 0, 0] [1, %arg1, 64] [1, 1, 1] : tensor<64x?x64xf32> to tensor<1x?x64xf32>
    %padded = tensor.pad %extracted_slice low[0, 0, 0] high[0, %0, 0] {
    ^bb0(%arg3: index, %arg4: index, %arg5: index):
      tensor.yield %arg0 : f32
    } : tensor<1x?x64xf32> to tensor<1x?x64xf32>
    return %padded : tensor<1x?x64xf32>
  }

  // mlir-opt --reify-result-shapes
  #map = affine_map<()[s0] -> (-s0 + 256)>
  func.func @func(%arg0: f32, %arg1: index, %arg2: tensor<64x?x64xf32>) -> tensor<1x?x64xf32> {
    %0 = affine.apply #map()[%arg1]
    %extracted_slice = tensor.extract_slice %arg2[0, 0, 0] [1, %arg1, 64] [1, 1, 1] : tensor<64x?x64xf32> to tensor<1x?x64xf32>
    %padded = tensor.pad %extracted_slice low[0, 0, 0] high[0, %0, 0] {
    ^bb0(%arg3: index, %arg4: index, %arg5: index):
      tensor.yield %arg0 : f32
    } : tensor<1x?x64xf32> to tensor<1x256x64xf32>
    %cast = tensor.cast %padded : tensor<1x256x64xf32> to tensor<1x?x64xf32>
    return %cast : tensor<1x?x64xf32>
  }
  ```

Commit:
[mlir][memref] Add a new InderStaticShapes pass for ReifyRankedShapedTypeOpInterface

fabianmcg committed:
rename transform to reify-result-shapes

MaheshRavishankar reviewed (changes requested):
This seems to be similar to https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/MemRef/Transforms/ResolveShapedTypeResultDims.cpp . What is this doing on top of that pass?

MaheshRavishankar commented:
Looking at 
```
#map = affine_map<(d0) -> (-d0 + 256)>
func.func @func(%arg0: f32, %arg1: index, %arg2: tensor<64x?x64xf32>) -> tensor<1x?x64xf32> {
  %0 = affine.apply #map(%arg1)
  %extracted_slice = tensor.extract_slice %arg2[0, 0, 0] [1, %arg1, 64] [1, 1, 1] : tensor<64x?x64xf32> to tensor<1x?x64xf32>
  %padded = tensor.pad %extracted_slice low[0, 0, 0] high[0, %0, 0] {
  ^bb0(%arg3: index, %arg4: index, %arg5: index):
    tensor.yield %arg0 : f32
  } : tensor<1x?x64xf32> to tensor<1x?x64xf32>
  return %padded : tensor<1x?x64xf32>
}

// mlir-opt --reify-result-shapes
#map = affine_map<()[s0] -> (-s0 + 256)>
func.func @func(%arg0: f32, %arg1: index, %arg2: tensor<64x?x64xf32>) -> tensor<1x?x64xf32> {
  %0 = affine.apply #map()[%arg1]
  %extracted_slice = tensor.extract_slice %arg2[0, 0, 0] [1, %arg1, 64] [1, 1, 1] : tensor<64x?x64xf32> to tensor<1x?x64xf32>
  %padded = tensor.pad %extracted_slice low[0, 0, 0] high[0, %0, 0] {
  ^bb0(%arg3: index, %arg4: index, %arg5: index):
    tensor.yield %arg0 : f32
  } : tensor<1x?x64xf32> to tensor<1x256x64xf32>
  %cast = tensor.cast %padded : tensor<1x256x64xf32> to tensor<1x?x64xf32>
  return %cast : tensor<1x?x64xf32>
}
```

This has nothing to do with reifying result shapes. This is just a const shape propagation that should be covered by the "cast folding" pattern here https://github.com/llvm/llvm-project/blob/d699fbd2033e291e8d658777d76b3b9ae1652f44/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp#L3394 (or the one below it).

There isnt a tensor.dim in the input IR, so there is nothing to "reify" . Might be a mismatch in what you expect that interface to do. It is explicitly meant to resolve `tensor.dim` operations.

fabianmcg commented:
We might need to update the docs, because at least I cannot infer that from the docs:
```
def ReifyRankedShapedTypeOpInterface :
    OpInterface<"ReifyRankedShapedTypeOpInterface"> {
  let description = [{
    Interface to compute the shape of the result of an operation when
    the result is a ranked shape type, i.e. `RankedTensorType` or
    `MemRefType`.
  }];
  let cppNamespace = "::mlir";

  let methods = [
    InterfaceMethod<
      /*desc=*/[{
        Reify the shape of the result of an operation (typically in terms of the
        shape of its operands).

        `reifiedReturnShapes` is populated with one vector per op result. Each
        of those vectors contains an OpFoldResult for each dimension of the
        shaped type. The given builder may be used to insert ops that compute
        result shapes.

        If the shape of a particular result cannot be computed it must be empty.
      }],
      /*retTy=*/"::llvm::LogicalResult",
      /*methodName=*/"reifyResultShapes",
      /*args=*/(ins "::mlir::OpBuilder &":$builder,
        "::mlir::ReifiedRankedShapedTypeDims &":$reifiedReturnShapes)
    >
  ];
}
```

Also, there's no cast to fold into or from in this case. So no way to augment those patterns.

Finally, I'd argue that ir is reification because we are coming from dynamic dimensions and inferring that the output type is static.

MaheshRavishankar commented:
> We might need to update the docs, because at least I cannot infer that from the docs:
> 
> ```
> def ReifyRankedShapedTypeOpInterface :
>     OpInterface<"ReifyRankedShapedTypeOpInterface"> {
>   let description = [{
>     Interface to compute the shape of the result of an operation when
>     the result is a ranked shape type, i.e. `RankedTensorType` or
>     `MemRefType`.
>   }];
>   let cppNamespace = "::mlir";
> 
>   let methods = [
>     InterfaceMethod<
>       /*desc=*/[{
>         Reify the shape of the result of an operation (typically in terms of the
>         shape of its operands).
> 
>         `reifiedReturnShapes` is populated with one vector per op result. Each
>         of those vectors contains an OpFoldResult for each dimension of the
>         shaped type. The given builder may be used to insert ops that compute
>         result shapes.
> 
>         If the shape of a particular result cannot be computed it must be empty.
>       }],
>       /*retTy=*/"::llvm::LogicalResult",
>       /*methodName=*/"reifyResultShapes",
>       /*args=*/(ins "::mlir::OpBuilder &":$builder,
>         "::mlir::ReifiedRankedShapedTypeDims &":$reifiedReturnShapes)
>     >
>   ];
> }
> ```
> 

Fair point. Document needs to be updated. 

> Also, there's no cast to fold into or from in this case. So no way to augment those patterns.
> 
Ok, maybe I misread what it the intent was and got thrown off by the use of the `ReifyRankedShapedTypeOpInterface`. (the reify was meant to say "generate the shape of the output" implicitly saying generate the code for `tensor.dim <result>, <dim>`.).

For other operations this is done by just canonicalizing the op itself to go to a more "static" version of the op, and introduce `tensor.cast` operations to fix up the tasks, and the `tensor.cast` folding patterns that are added absorb these into producer/consumers and propagate the "static" information through the program. AFAICS, the first pattern is missing and that has nothing to do with the `ReifyRankedShapedTypeOpInterface`. We need to add a pattern by itself that is similar to what is done for `tensor.extract_slice` [here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp#L2724). Sorry I think you intended to do that from the get go, but I was thrown off by the `ReifyRankedShapedTypeOpInterface` angle.

> Finally, I'd argue that it is reification because we are coming from dynamic dimensions and inferring that the output type is static.

nicolasvasilache commented:
> Ok, maybe I misread what it the intent was and got thrown off by the use of the `ReifyRankedShapedTypeOpInterface`. (the reify was meant to say "generate the shape of the output" implicitly saying generate the code for `tensor.dim <result>, <dim>`.).

There is no need to limit the interface to only exist in conjunction with `tensor.dim` ops, see my explanation here https://github.com/llvm/llvm-project/pull/145732#issuecomment-3012048686

I also don't understand "implicitly saying generate the code for `tensor.dim <result>, <dim>`": this is really a shape question, which may or may not involve `tensor.dim` ops. The more we fold away such ops the better to obtain static shapes.

Happy to rename the functionality if that is what causes hiccups (naming is still hard in 2025).

Commit:
Don't fail the pass if we can't make shapes more static.

Also, only allow tensor::PadOp and tensor::ConcatOp for now as more extensive testing showed that
other ops are not ready yet (e.g. at least tensor::ExtractSliceOp / tensor::InsertSliceOp).

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Transforms/ReifyResultShapes.cpp:
```diff
@@ -0,0 +1,148 @@
+  LLVM_DEBUG({ DBGS() << " reifying op: " << op << "\n"; });
+  // Get the reified out shapes.
+  ReifiedRankedShapedTypeDims reifiedResultShapes;
+  if (failed(mlir::reifyResultShapes(rewriter, op, reifiedResultShapes)) ||
```
> To me the interface description establishes an implicit contract allowing this:
> 
> ```
>     Interface to compute the shape of the result of an operation when
>     the result is a ranked shape type, i.e. `RankedTensorType` or
>     `MemRefType`.
> ```
> 
> Because, what would it mean for `reifyResultShapes` to return a shape that the op verifier will reject? The interface would produce inconsistent results with itself, rendering the interface implementation erroneous (IMO the verifier has higher precedence).
> 

I am not sure I fully follow the logic. Lets assume the op is currently valid and you are getting the shape of the result. Now you are change the shape of the result without modifying any of its other operands. There is no way you can make an interface gaurantee that the shape it found is considered valid by the operation. The operation method itself might not know that. The inconsistency could be coming from the verifier checking consistency between the other operands values and its result type. You really cannot expect the `reify` method implementation to also know whether the op is valid or not, that is kind of a disconnected contract that can be a source of bugs.

Commit:
Update pass documentation

MaheshRavishankar commented:
I looked through the original "canonicalization" PR <https://github.com/llvm/llvm-project/pull/145732> . This has the same issue that I am raising here. The `reifyResultShape` has more information about the op shapes than the op itself. That cannot be the case (we can all talk about not having enough documentation of things, and for sure lets fix documentation whenever something is not clear).

The right ordering here would be a pattern that would do what is done in the `reify` method here  https://github.com/llvm/llvm-project/blob/01d0171e7a955deff032bed3ac1b7e51d98ce4a9/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp#L3811  (that could be run as part of canonicalization), and the `reify` method just gets the information from the operation. The operation has to be the source of information here. 

If you want me to stamp this as a WIP/to-be-evolved thing, thats fine, but it is relying on what I see as an inverted dependence structure between the op and the interface method. I would have raised this on the implementation of the `reifyResultShape` to begin with if it hadnt sent/reviewed/landed before anyone from a different time zone had any chance to review it.

nicolasvasilache commented:
> and the reify method just gets the information from the operation. The operation has to be the source of information here.

Documenting offline discussion: for a meaningful class of ops, the "transfer function" can be encoded with `IndexingOpInterface`. In the fullness of time, this would likely merge with part of the `ReifyRankedShapedTypeOpInterface` interface to become "the source of truth for the op" semantics. 
There are extra considerations wrt building IR and taking a builder or not:
- canonicalization can't build IR in the case of failure
- verifier can't even take a builder
- reify freely takes a builder and creates IR
For the class of things that can be constructed with an AffineMap transfer function, all these things can and should agree with each other, achieving confluence.

For other things, TBD what other meaningful classes of ops can have "transfer functions" that reconcile the SSA-less builder-less verifier with the SSA + builder + folding `ReifyRankedShapedTypeOpInterface`.

Commit:
Review comments
